lib/elixir/lib/map_set.ex

```diff
@@ -226,7 +231,7 @@ defmodule MapSet do
       #MapSet<[]>
 
   """
-  @spec intersection(t, t) :: t
+  @spec intersection(t(val1), t(val2)) :: t(val1 | val2) when val1: value, val2: value
```

Comment:
Good point, I got what you mean @fishcakez and yes we can't describe this, so let's go with `val`.